Add Monitoring V2

aiyengar2 opened the pull request:
Introduces two charts:
- `rancher-monitoring`: deploys Prometheus Operator + Prometheus Adapter + a set of PushProx exporters depending on the k8s distribution the user is using (i.e. RKE, k3s, kubeAdm, etc.) 
- `rancher-pushprox`: deploys a single PushProx exporter consisting of a Deployment + Service of a proxy, a DaemonSet + Service of clients, a ServiceMonitor, and some RBAC components (if using a service account to make authorized scrapes on k8s components)

All the changes from upstream Prometheus Operator have been documented in the `overlay/CHANGELOG.md` file that can be viewed in each commit to see the specific changes that each commit introduces to the chart.

aiyengar2 commented on docs/index.yaml:
```diff
@@ -3,7 +3,7 @@ entries:
   - apiVersion: v1
     appVersion: v3.13.3
-    created: "2020-06-22T19:23:35.619402732Z"
+    created: "2020-06-26T11:06:00.318061-07:00"
```
@StrongMonkey it seems like all of the timestamps in`docs/index.yaml` are being updated on running `make CHART=rancher-monitoring-v2 charts` due to https://github.com/rancher/dev-charts/blob/f091a152fbe8ece1a05a4cb8e800aa217d48f825/scripts/generate-charts#L22

Is this expected?

StrongMonkey replied:
Yes, I think this is a known issue. Don't know how to fix it yet.

StrongMonkey commented:
@aiyengar2 There is no need to commit changes for docs/index.yaml and tar.gz files. CI will publish that. You will just need to commit patch file and package.yaml.

aiyengar2 commented:
> @aiyengar2 There is no need to commit changes for docs/index.yaml and tar.gz files. CI will publish that. You will just need to commit patch file and package.yaml.

Sounds good, removed those files from the commit!

Commit:
Add rancher-pushprox chart

Introduces a new chart called rancher-pushprox that deploys:
- A Deployment of a PushProx proxy + Service
- A DaemonSet of PushProx clients + Service
- A ServiceMonitor
- RBAC components for setting up authorized scraping

Accordingly, the ServiceMonitor will scrape client metrics through the proxy.

Using an instance of this chart is suitable for the following scenarios:
- You need to scrape metrics from a port that should not be accessible outside of the host (e.g. scraping `etcd` metrics in a hardened cluster)
- You need to scrape metrics on a host that are not exposed outside of 127.0.0.1 (e.g. scraping `kube-proxy` metrics)
- You need to scrape metrics through HTTPS using certs hosted directly on `hostPath`
- You need to scrape metrics from Kubernetes components that require authorization via a service account (e.g. permissions to make request to `/metrics`)
- You need to scrape metrics without access to cacerts (i.e. enable `insecureSkipVerify`)

The clients can be configured to:
- Scrape metrics on the hostNetwork of the node it is deployed on at a particular port
- Scrape metrics over HTTPS using ServiceAccount credentials
- Scrape metrics over HTTPS using insecureSkipVerify
- Scrape metrics over HTTPS using TLS certs, keys and cacerts within a hostPath
- Deploy on specific nodes based on nodeSelectors and tolerations

The proxy can also be configured to deploy on specific nodes based on nodeSelectors and tolerations.

See README.md for more details.

This chart will be used by the rancher-monitoring chart to set up component monitoring across k8s distributions.

Commit:
Add rancher-monitoring

This commit creates a Monitoring V2 chart from the upstream Prometheus Operator Helm chart.

Prometheus Operator: https://github.com/helm/charts/tree/master/stable/prometheus-operator

See changes to `overlay/CHANGELOG.md` for details on what has been added/modified.

Commit:
Add default ClusterRoles for monitoring

Added the following default ClusterRoles (see CHANGELOG.md for more details):
- `monitoring-admin` (aggregates to default admin)
- `monitoring-edit` (aggregates to default edit)
- `monitoring-view` (aggregates to default view)
- `grafana-config-edit`
- `grafana-config-view`

See changes to `overlay/CHANGELOG.md` for details on what has been added/modified.

Commit:
Provide separate namespaces for Grafana ConfigMaps

This commit makes a couple of changes to support Grafana ConfigMaps:
- Exposes the `searchNamespace` field of the Grafana datasource sidecar
- Exposes the `searchNamespace` field of the Grafana dashboard sidecar
- Adds `Namespace` resources for the `searchNamespaces` specified
- Moves all Grafana ConfigMaps to the corresponding namespaces

See changes to `overlay/CHANGELOG.md` for details on what has been added/modified.

Commit:
Add RKE support to rancher-monitoring-v2

Adds support for monitoring the following k8s components in RKE clusters:
- `kube-scheduler`
- `kube-controller-manager`
- `kube-proxy`
- `kube-etcd`

All exporters are created from the cattle-pushprox chart. This commit
also modifies the relevant Grafana Dashboard ConfigMaps and
PrometheusRules to deploy if the PushProx exporters are enabled.

See changes to `overlay/CHANGELOG.md` for details on what has been added/modified.

Commit:
Add k3s support to rancher-monitoring-v2

Adds support for monitoring the following k8s components in k3s clusters:
- `kube-scheduler`
- `kube-controller-manager`
- `kube-proxy`

All exporters are created from the cattle-pushprox chart. This commit
also modifies the relevant Grafana Dashboard ConfigMaps and
PrometheusRules to deploy if the PushProx exporters are enabled.

See changes to `overlay/CHANGELOG.md` for details on what has been added/modified.

Commit:
Add kubeAdm support to rancher-monitoring-v2

Adds support for monitoring the following k8s components in kubeAdm clusters:
- `kube-scheduler`
- `kube-controller-manager`
- `kube-proxy`
- `kube-etcd`

All exporters are created from the cattle-pushprox chart. This commit
also modifies the relevant Grafana Dashboard ConfigMaps and
PrometheusRules to deploy if the PushProx exporters are enabled.

See changes to `overlay/CHANGELOG.md` for details on what has been added/modified.

deniseschannon commented on packages/cattle-pushprox/README.md:
```diff
@@ -0,0 +1,52 @@
+The clients and proxy are created based on of a Rancher fork of the Prometheus Community [PushProx](https://github.com/prometheus-community/PushProx) project.
+
+Using an instance of this chart is suitable for the following scenarios:
+- You need to scrape metrics from a port that should not be accessible outside of the host (i.e. scraping `etcd` metrics in a hardened cluster)
```
Small nit, but these should be `e.g.`  (for example) not `i.e.` (that is)

aiyengar2 replied:
good catch, updated in latest push

pennyscissors reviewed (dismissed):
lgtm

deniseschannon commented on packages/cattle-pushprox/Chart.yaml:
```diff
@@ -0,0 +1,23 @@
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+appVersion: 1.16.0
```
Add the Rancher certified annotation to the chart. 

```
annotations:
  io.rancher.certified: rancher
```

aiyengar2 replied:
added to the latest push!

dramich commented on packages/cattle-pushprox/Chart.yaml:
```diff
@@ -0,0 +1,25 @@
+apiVersion: v1
+name: cattle-pushprox
+description: A Rancher chart based on PushProx that sets up a deployment of the PushProx proxy and several PushProx clients on a Kubernetes cluster.
```
Can this be clarified? What does 'several' mean?

aiyengar2 replied:
Sure, will change it to `a DaemonSet of PushProx clients`

dramich commented on packages/cattle-pushprox/README.md:
```diff
@@ -0,0 +1,52 @@
+
+Installs [cattle-pushprox](https://github.com/rancher/dev-charts/tree/master/packages/cattle-pushprox) to create a DaemonSet of clients that can access their host's network and register with a PushProx proxy Deployment. A [Prometheus Operator](https://github.com/coreos/prometheus-operator) ServiceMonitor CR is also included that is configured to scrape the metrics from each of the clients through ...
+
+The clients and proxy are created based on of a Rancher fork of the Prometheus Community [PushProx](https://github.com/prometheus-community/PushProx) project.
```
```suggestion
The clients and proxy are created based on a Rancher fork of the Prometheus Community [PushProx](https://github.com/prometheus-community/PushProx) project.
```

aiyengar2 replied:
Good catch, will be updated in the latest commit

dramich commented on packages/cattle-pushprox/templates/pushprox-clients.yaml:
```diff
@@ -0,0 +1,159 @@
+      {{- end }}
+      containers:
+      - name: pushprox-client
+        image: arvindiyengar/pushprox-linux-amd64:add_flag_for_token_path
```
Should be a rancher hosted image.

aiyengar2 replied:
This is pending merges from https://github.com/rancher/PushProx/pull/1 and https://github.com/rancher/PushProx/pull/2 for initializing the Rancher repo where the hosted image will be generated from

dramich commented on packages/cattle-pushprox/templates/pushprox-clients.yaml:
```diff
@@ -0,0 +1,159 @@
+          KEY_FILE_SOURCE=$(find /etc/source/ -type f -name "${KEY_FILE_NAME}" | sort -r | head -n 1)
+          CACERT_FILE_SOURCE=$(find /etc/source/ -type f -name "${CACERT_FILE_NAME}" | sort -r | head -n 1)
+          
+          test -z ${CERT_FILE_SOURCE} && echo "Failed to find cacert file" && exit 1
+          test -z ${CERT_FILE_SOURCE} && echo "Failed to find cert file" && exit 1
+          test -z ${CERT_FILE_SOURCE} && echo "Failed to find key file" && exit 1
```
These all do the same thing

aiyengar2 replied:
Good catch, this should be:
```suggestion
          test -z ${CERT_FILE_SOURCE} && echo "Failed to find cert file" && exit 1
          test -z ${KEY_FILE_SOURCE} && echo "Failed to find key file" && exit 1
          test -z ${CACERT_FILE_SOURCE} && echo "Failed to find cacert file" && exit 1
```

Making the change in the next commit

deniseschannon commented on packages/rancher-monitoring/rancher-monitoring.patch:
```diff
@@ -0,0 +1,1352 @@
+ appVersion: 0.38.1
++annotations:
++  catalog.rancher.io/certified: rancher
++  catalog.rancher.io/namespace: rancher-monitoring
```
Let's update this to `monitoring-system`. UI has requested that all the namespaces end in `-system` to allow filtering.

aiyengar2 replied:
Updated in the latest commit

deniseschannon commented on packages/rancher-pushprox/Chart.yaml:
```diff
@@ -0,0 +1,23 @@
+apiVersion: v1
+appVersion: 1.16.0
+annotations:
```
Unclear if these are needed at this level, but let's leave them in for now.

aiyengar2 replied:
I updated the appVersion to 0.1.0 to match the version of the upstream PushProx we are deploying:

https://github.com/rancher/PushProx/blob/753361c573410dc535028be56022962b9df3b9a6/VERSION#L1

deniseschannon commented on packages/rancher-monitoring/rancher-monitoring.patch:
```diff
@@ -0,0 +1,1352 @@
++  catalog.rancher.io/certified: rancher
++  catalog.rancher.io/namespace: monitoring-system
++  catalog.rancher.io/release-name: rancher-monitoring
+ description: Provides easy monitoring definitions for Kubernetes services, and deployment
```
We should update the description to add in what other charts that we're adding as part of our version of the chart.

aiyengar2 replied:
Updated to `A Rancher chart that modifies the upstream Prometheus Operator chart, which provides easy monitoring definitions for Kubernetes services and the deployment and management of Prometheus instances, and enables Prometheus Adapter on a default Prometheus instance.`

deniseschannon reviewed (approved):
LGTM for packaging
